### PR TITLE
Implement edition 2024 match ergonomics restrictions

### PR DESCRIPTION
**Pull request by Nadrieril**
This implements the minimalest version of [match ergonomics for edition 2024](https://rust-lang.github.io/rfcs/3627-match-ergonomics-2024.html). This minimal version makes it an error to ever reset the default binding mode. The implemented proposal is described precisely [here](https://hackmd.io/zUqs2ISNQ0Wrnxsa9nhD0Q#RFC-3627-nano), where it is called "RFC 3627-nano".

Rules:
- Rule 1C: When the DBM (default binding mode) is not `move` (whether or not behind a reference), writing `mut`, `ref`, or `ref mut` on a binding is an error.
- Rule 2C: Reference patterns can only match against references in the scrutinee when the DBM is `move`.

This minimal version is forward-compatible with the main proposals for match ergonomics 2024: [RFC3627](https://rust-lang.github.io/rfcs/3627-match-ergonomics-2024.html) itself, the alternative [rule 4-early variant](https://rust-lang.github.io/rfcs/3627-match-ergonomics-2024.html), and [others](https://hackmd.io/zUqs2ISNQ0Wrnxsa9nhD0Q). The idea is to give us more time to iron out a final proposal.

This includes a migration lint that desugars any offending pattern into one that doesn't make use of match ergonomics. Such patterns have identical meaning across editions.

This PR insta-stabilizes the proposed behavior onto edition 2024.

r? @ghost

Tracking:

- https://github.com/rust-lang/rust/issues/123076

### CODE REVIEW

**Comment by Nadrieril** [comment]
r? compiler

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by tmandry** [comment]
@rfcbot reviewed

**Comment by pnkfelix** [comment]
@rfcbot reviewed

File: tests/ui/pattern/rfc-3627-match-ergonomics-2024/migration_lint.stderr [hunk 1]
```diff
@@ -0,0 +1,188 @@
+error: patterns are not allowed to reset the default binding mode in edition 2024
+  --> $DIR/migration_lint.rs:25:9
+   |
+LL |     let Foo(mut x) = &Foo(0);
+   |         -^^^^^^^^^
+   |         |
+   |         help: desugar the match ergonomics: `&`
```
Comment (pnkfelix) [inline tests/ui/pattern/rfc-3627-match-ergonomics-2024/migration_lint.stderr:7]: I think this message may be a little bit too obtuse for the average reader.

In particular, my guess is that the average person leveraging match ergnonomics may not even *know* that they are doing so, and so they will not necessarily understand what this help message is telling them to do.

I recommend rephrasing it with a little more direction. Perhaps: "add a `&` to the pattern you have written to correspond to the type of the expression being matched" ?

**Comment by pnkfelix** [comment]
I had a [nit regarding the diagnostic messaging here](https://github.com/rust-lang/rust/pull/131381#pullrequestreview-2372680926), but I don't think that needs to block landing this PR.

File: tests/ui/pattern/rfc-3627-match-ergonomics-2024/migration_lint.stderr (hunk 1 above)
Comment (Nadrieril) [reply tests/ui/pattern/rfc-3627-match-ergonomics-2024/migration_lint.stderr:7]: oh yeah excellent point

**Pull request merged**